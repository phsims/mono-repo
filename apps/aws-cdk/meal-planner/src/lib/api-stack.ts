// import { Stack, RemovalPolicy } from 'aws-cdk-lib';

// import { Construct } from 'constructs';

// export class DynamoDbStack extends Stack {
//   public readonly table: Table;

//   constructor(scope: Construct, id: string) {
//     super(scope, id);

//   }
// }
