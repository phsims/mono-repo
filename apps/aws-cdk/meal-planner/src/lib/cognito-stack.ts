import { Stack, StackProps, RemovalPolicy } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { UserPool, UserPoolClient } from 'aws-cdk-lib/aws-cognito';

export class CognitoStack extends Stack {
  public readonly userPool: UserPool;
  public readonly userPoolClient: UserPoolClient;

  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    // Cognito Pool
    const userPool = new UserPool(this, 'userPool', {
      userPoolName: `${id}-pool`,
      removalPolicy: RemovalPolicy.DESTROY,
    });
    this.userPool = userPool;

    // App Client
    const userPoolClient = new UserPoolClient(this, 'userPoolClient', {
      userPool: userPool,
      userPoolClientName: `${id}-client`,
      authFlows: {
        userPassword: true,
      },
    });
    this.userPoolClient = userPoolClient;
  }
}
