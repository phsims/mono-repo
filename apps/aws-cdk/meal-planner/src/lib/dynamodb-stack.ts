import { Stack, RemovalPolicy } from 'aws-cdk-lib';
import {
  AttributeType,
  GlobalSecondaryIndexProps,
  Table,
  BillingMode,
} from 'aws-cdk-lib/aws-dynamodb';
import { Construct } from 'constructs';

export class DynamoDbStack extends Stack {
  public readonly table: Table;

  constructor(scope: Construct, id: string) {
    super(scope, id);

    //setup table
    const table = new Table(this, `${id}-table`, {
      partitionKey: { name: 'PK', type: AttributeType.STRING },
      billingMode: BillingMode.PAY_PER_REQUEST,
      removalPolicy: RemovalPolicy.DESTROY,
      sortKey: { name: 'SK', type: AttributeType.STRING },
    });
    this.table = table;

    // setup indexs
    const gsi1: GlobalSecondaryIndexProps = {
      indexName: 'GSI1',
      partitionKey: {
        name: 'GSI1PK',
        type: AttributeType.STRING,
      },
      sortKey: {
        name: 'GSI1SK',
        type: AttributeType.STRING,
      },
    };
    table.addGlobalSecondaryIndex(gsi1);

    const gsi2: GlobalSecondaryIndexProps = {
      indexName: 'GSI2',
      partitionKey: {
        name: 'GSI2PK',
        type: AttributeType.STRING,
      },
      sortKey: {
        name: 'GSI2SK',
        type: AttributeType.NUMBER,
      },
    };

    table.addGlobalSecondaryIndex(gsi2);
  }
}
