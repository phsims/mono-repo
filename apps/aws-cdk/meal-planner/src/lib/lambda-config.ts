import { Duration } from 'aws-cdk-lib';
import {
  LambdaDefinition,
  LambdaDefinitionProps,
  LambdaFunctionProps,
} from '../lambda-layer/types';
import { NodejsFunctionProps } from 'aws-cdk-lib/aws-lambda-nodejs';
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import * as path from 'path';
import { UserPool } from 'aws-cdk-lib/aws-cognito';
import { HttpMethod } from 'aws-cdk-lib/aws-stepfunctions-tasks';

// Returns lambda definitions with custom env
export const getLambdaDefinitions = (
  id: string,
  userPool?: UserPool
): LambdaDefinition[] => {
  const environment = {
    DDB_TABLE: `${id}`,
  };
  const lambdaDefinitions: LambdaDefinition[] = [
    // {
    //   name: 'get-users',
    //   environment,
    //   api: {
    //     path: '/users',
    //     methods: [HttpMethod.GET],
    //   },
    // },
    {
      name: 'add-user',
      environment,
      api: {
        path: '/users',
        methods: [HttpMethod.POST],
      },
    },
    // {
    //   name: 'delete-user',
    //   environment,
    //   api: {
    //     path: '/users',
    //     methods: [HttpMethod.DELETE],
    //   },
    // },
    // {
    //   name: 'update-user',
    //   environment,
    //   api: {
    //     path: '/users',
    //     methods: [HttpMethod.PATCH],
    //   },
    // },
    // {
    //   name: 'api-authorizer',
    //   environment: { ...environment, USER_POOL_ID: userPool.userPoolId },
    // },
  ];
  return lambdaDefinitions;
};

// Returns Lambda Function properties with defaults and overwrites
export const getFunctionProps = ({
  id,
  lambdaDefinition,
  lambdaRole,
  lambdaLayer,
}: LambdaFunctionProps): NodejsFunctionProps => {
  const functionProps: NodejsFunctionProps = {
    functionName: `${id}-${lambdaDefinition.name}`,
    entry: path.join(
      __dirname,
      '..',
      'lambda-handlers',
      `${lambdaDefinition.name}.ts`
    ),
    runtime: Runtime.NODEJS_18_X,
    role: lambdaRole,
    layers: [lambdaLayer],
  };
  return functionProps;
};
