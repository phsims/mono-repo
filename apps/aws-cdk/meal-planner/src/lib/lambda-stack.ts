import { Stack, StackProps, RemovalPolicy } from 'aws-cdk-lib';
import { Runtime, LayerVersion, Code } from 'aws-cdk-lib/aws-lambda';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { Construct } from 'constructs';
import { Role, ServicePrincipal, ManagedPolicy } from 'aws-cdk-lib/aws-iam';
import * as path from 'path';

import { LambdaStackProps } from '../lambda-layer/types';

export class LambdaStack extends Stack {
  constructor(scope: Construct, id: string, props: LambdaStackProps) {
    super(scope, id, props);

    // Lambda Role
    const lambdaRole = new Role(this, 'lambdaRole', {
      roleName: `${id}-lambda-role`,
      description: `Lambda role for ${id}`,
      assumedBy: new ServicePrincipal('amazonaws.com'),
      managedPolicies: [
        ManagedPolicy.fromAwsManagedPolicyName('ReadOnlyAccess'),
      ],
    });

    // Add DDB read write permissions to lambda role
    props.table.grantReadWriteData(lambdaRole);

    // Lambda Layer
    const lambdaLayer = new LayerVersion(this, 'lambdaLayer', {
      code: Code.fromAsset(path.join(__dirname, '../lambda-layer')),
      compatibleRuntimes: [Runtime.NODEJS_18_X],
      description: `Lambda Layer for ${id}`,
    });

    // lambda function

    // const dynamoLambda = new Function(this, "DynamoLambdaHandler", {
    //   runtime: Runtime.NODEJS_18_X,
    //   code: Code.asset("functions"),
    //   handler: "function.handler",
    //   environment: {
    //     HELLO_TABLE_NAME: props.table.tableName,
    //   },
    // });
  }
}
