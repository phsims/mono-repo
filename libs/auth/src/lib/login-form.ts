export class LoginForm extends HTMLElement {
  connectedCallback() {
    this.innerHTML = `
    <div>
    <input type="email" />
    <input type="password" />
    <button/>
    </div>`;
  }
}

customElements.define('login-form', LoginForm);
