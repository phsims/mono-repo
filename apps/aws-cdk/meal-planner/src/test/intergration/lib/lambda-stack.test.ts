import { Stack, App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { DynamoDbStack } from '../../../lib/dynamodb-stack';
import { CognitoStack } from '../../../lib/cognito-stack';
import { LambdaStack } from '../../../lib/lambda-stack';

test('LambdaStack Stack', () => {
  const app = new App();
  const dynamoDbStack = new DynamoDbStack(app, `TestStack-dynamoDb-stack`);
  const cognitoStack = new CognitoStack(app, `TestStack-cognito-stack`);

  const stack = new LambdaStack(app, 'TestStack', {
    table: dynamoDbStack.table,
    userPool: cognitoStack.userPool,
  });
  const template = Template.fromStack(stack as Stack);

  expect(template).toMatchSnapshot('TestStack');
});
