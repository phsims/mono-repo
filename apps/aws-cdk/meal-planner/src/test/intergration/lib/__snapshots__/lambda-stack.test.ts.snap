// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LambdaStack Stack: TestStack 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "lambdaLayer857D4AF4": {
      "Properties": {
        "CompatibleRuntimes": [
          "nodejs18.x",
        ],
        "Content": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "ad46ff9b56ae47ed45e6fb05f78c0386a972f2e224df7956b255e9683881a0b2.zip",
        },
        "Description": "Lambda Layer for TestStack",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "lambdaRoleC844FDB1": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Lambda role for TestStack",
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/ReadOnlyAccess",
              ],
            ],
          },
        ],
        "RoleName": "TestStack-lambda-role",
      },
      "Type": "AWS::IAM::Role",
    },
    "lambdaRoleDefaultPolicyA63A8A92": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::ImportValue": "TestStack-dynamoDb-stack:ExportsOutputFnGetAttTestStackdynamoDbstacktableB0D65BCCArnC5CBA71C",
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::ImportValue": "TestStack-dynamoDb-stack:ExportsOutputFnGetAttTestStackdynamoDbstacktableB0D65BCCArnC5CBA71C",
                      },
                      "/index/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "lambdaRoleDefaultPolicyA63A8A92",
        "Roles": [
          {
            "Ref": "lambdaRoleC844FDB1",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
